<launch>
    <arg name="forkpos" default="/forklift_pose"/>
    <node pkg="tf" type="static_transform_publisher" name="link1_broadcaster" args="0.0 0.22 0.43 1.57 0.0 0 base_link camera_link 100" />

    <!--megapose-->
    <!-- <include file="$(find visp_megapose)/launch/megapose_client_oilpalm.launch"/> -->

    <node pkg="forklift_server" type="PBVS_server_differential.py" name="PBVS_server" output="screen">
        <!--Subscriber Topic setting/-->
        <param name="arm_ID" value="2" /><!--剪鉗1 or 2/-->
        <param name="odom" value="/wheel_odom" /><!--里程計 Topic（map -> base_link, 是tf相對關係, 非輪式里程計）/-->
        <param name="pose_topic" value="/apple" /><!--相機對位的 Topic/-->
        <param name="object_filter" value="False" /><!--(True: , False: )/-->
        <param name="confidence_minimum" value="0.4" /><!--foundationpose信心值低於它停止動作/-->
        <param name="arm_status_topic" value="/arm_current_status" /><!--剪鉗狀態 Topic/-->
        <param name="arm_control_topic" value="/cmd_cut_pliers_2" /><!--剪鉗動作 Topic/-->

        <!--camera parking setting-->
        <param name="camera_side" value="right"/><!--left or right/-->
        <param name="camera_tag_offset_x" type="double" value = "0.0" /><!--對位目標點與tag的左右偏移量/-->
        <param name="camera_desired_dist_threshold" type="double" value = "0.4" />  <!--距離目標點的直線距離小於threshold，就不執行move_nearby_parking_lot動作-->
        <param name="camera_horizon_alignment_threshold" type="double" value = "0.03" />  <!--車體相機水平對齊偏差容許距離（公尺）/-->
    
        <!--Cut pliers (arm) control settings/-->
        <param name="spin_forward_comp" type="double" value="0.02"/><!-- 旋轉造成的前向補償 (m)，相機在左/右都適用 -->
        
        <!-- Anti-overshoot（mm） -->
        <param name="marker_to_mm" value="1000.0"/><!-- 相機 Pose 單位轉換：若 Pose 為 m，設 1000；若 Pose 已是 mm，設 1 -->
        
        <param name="z_lead_near_mm" value="6.0"/><!-- 預停距離：當前目標與預停點距離小於此值，直接移動到目標 -->
        <param name="z_lead_far_mm" value="12.0"/><!-- 預停距離：當前目標與預停點距離大於此值，使用遠距離預停距離 -->
        <param name="z_lead_switch_mm" value="10.0"/><!-- 預停距離切換點：當前目標與預停點距離大於此值，使用遠距離預停距離 -->
        <param name="z_step_max_mm" value="15.0"/><!-- 最大步進距離：每次移動的最大距離 -->

        <param name="x_lead_near_mm" value="10.0"/><!-- 預停距離：當前目標與預停點距離小於此值，直接移動到目標 -->
        <param name="x_lead_far_mm" value="20.0"/><!-- 預停距離：當前目標與預停點距離大於此值，使用遠距離預停距離 -->
        <param name="x_lead_switch_mm" value="20.0"/><!-- 預停距離切換點：當前目標與預停點距離大於此值，使用遠距離預停距離 -->
        <param name="x_step_max_mm" value="25.0"/><!-- 最大步進距離：每次移動的最大距離 -->

        <param name="zx_stable_frames" value="5"/>

        <!-- 安全行程（mm） -->
        <param name="cut_pliers_min_height" value="0.0"/><!--手臂最小高度（用於 fnControlArmBasedOnFruitZ）/-->
        <param name="cut_pliers_max_height" value="280.0"/><!--手臂最大高度（用於 fnControlArmBasedOnFruitZ）/-->
        <param name="cut_pliers_min_length" value="10.0" /><!--手臂最小長度（用於 fnControlArmBasedOnFruitX 和 fnBlindExtendArm）/-->
        <param name="cut_pliers_max_length" value="280.0"/><!--手臂最大長度（用於 fnControlArmBasedOnFruitX 和 fnBlindExtendArm）/-->
        
        <param name="z_tolerance_mm" value="5.0" /><!--Z 軸容許誤差（用於 fnControlArmBasedOnFruitZ）/-->
        <param name="z_step_min_mm" value="2.0" /><!--手臂高度調整最小步長（用於 fnControlArmBasedOnFruitZ）/-->
        <param name="z_step_max_mm" value="15.0" /><!--手臂高度調整最大步長（用於 fnControlArmBasedOnFruitZ）/-->
        <param name="z_step_gain" value="0.5" /><!--手臂高度調整步長增益（用於 fnControlArmBasedOnFruitZ）/-->

        <param name="cut_pliers_allow_retract" value="false" /><!--是否允許手臂後退（用於 fnControlArmBasedOnFruitX）/-->
        <param name="x_tolerance_mm" value="5.0" /><!--X 軸容許誤差（用於 fnControlArmBasedOnFruitX）/-->
        <param name="x_step_min_mm" value="2.0" /><!--手臂長度調整最小步長（用於 fnControlArmBasedOnFruitX）/-->
        <param name="x_step_max_mm" value="15.0" /><!--手臂長度調整最大步長（用於 fnControlArmBasedOnFruitX）/-->
        <param name="x_step_gain" value="0.5" /><!--手臂長度調整步長增益（用於 fnControlArmBasedOnFruitX）/-->
        <param name="x_cut_pliers_target" type="double" value="0.15" /><!--水果 X 軸目標值（用於 fnControlArmBasedOnFruitX）/-->
        <param name="cut_pliers_blind_extend_length" type="double" value="150.0" /><!--盲伸的額外長度（用於 DeadMoveX）/-->
        <param name="cut_pliers_home_length" type="double" value="10.0" /><!--手臂回原長度（GoHome）/-->
        <param name="cut_pliers_home_height" type="double" value="140.0" /><!--手臂回原高度（GoHome）/-->
    </node>
</launch>